RULES/cc/cc.go

```diff
@@ -118,41 +123,10 @@ type Library struct {
 	Shared        bool
 	AlwaysLink    bool
 	Toolchain     Toolchain
-}
-
-// multipleToolchainLibrary is a library that can be built
-// via multiple different toolchains from the same build.
-type multipleToolchainLibrary struct {
-	baseOut core.OutPath
-	lib     Library
-}
-
-func (lib Library) MultipleToolchains() multipleToolchainLibrary {
-	if lib.Out == nil {
-		core.Fatal("Out field is required for cc.Library")
-	}
-	return multipleToolchainLibrary{
-		baseOut: lib.Out,
-		lib:     lib,
-	}
-}
-
-// CcLibrary for a multi-toolchain-library returns a toolchain-specific
-// copy of the registered library.
-func (mtl multipleToolchainLibrary) CcLibrary(toolchain Toolchain) Library {
-	toolchain = toolchainOrDefault(toolchain)
-	tcLib := mtl.lib
-	if toolchain.Name() != DefaultToolchain().Name() {
-		tcLib.Out = mtl.baseOut.WithPrefix(toolchain.Name() + "/")
-	}
-	tcLib.Toolchain = toolchain
-	return tcLib
-}
 
-// Build for a multi-toolchain-library builds the library
-// with the default toolchain.
-func (mtl multipleToolchainLibrary) Build(ctx core.Context) {
-	mtl.CcLibrary(DefaultToolchain()).Build(ctx)
+	// Extra fields for handling multi-toolchain logic.
+	userOut       core.OutPath
```

Comment:
Are these two fields (userOut and userToolchain) needed?

What happens if you just set Out and Toolchain of the returned Library from CcLibrary? (Note that everything in these rules are structs and not pointers, so the original library is untouched).

Reply:
Yeah I wanted to ensure that calling `CcLibrary()` multiple times creates sensible results. For example `libA.CcLibrary('A').CcLibrary('B')` would just blindly prefix the `lib.Out` path with both `A` and `B` toolchains without `userOut` and `userToolchain`. This is because during the second `CcLibrary()` call there's no way to tell what was the original `lib.Toolchain` and `lib.Out`.

Reply:
Perhaps it's better to make that an error? Would it ever be meaningful to call CcLibrary more than once?

There's something iffy about this code, which I think is because Library is used for two things -- specifying a library in a BUILD file, but also as a lower-level thing which is a dependency of a cc target. I know this isn't introduced by this pull request, and I'm not suggesting that you necessarily fix it.